src/tracedAtomic.ml

```diff
@@ -109,167 +125,298 @@ let handler current_process_id runner =
+  let j = proc.proc_id in
+  let find ptr map = match IdMap.find_opt ptr map with
+    | None -> None
+    | Some lst ->
+        List.find_opt (fun (_, proc_id) -> proc_id <> j) lst
+  in
+  let find_loc proc =
+    match categorize proc.op, proc.obj_ptr with
+    | Ignore, _ -> None
+    | Read, Some ptr -> find ptr last_write
+    | (Write | Read_write), Some ptr -> max (find ptr last_read) (find ptr last_write)
+    | _ -> assert false
+  in
+  let rec find_replay_trace ~lower ~upper proc_id =
+    let pre_s = List.nth state (time - upper + 1) in
+    let replay_steps = List.filteri (fun k s -> k >= lower && k <= time - upper && s.run.proc_id = proc_id) state in
+    let replay_steps = List.rev_map (fun s -> s.run) replay_steps in
+    let causal p = match find_loc p with
+      | None -> true
+      | Some (k, _) -> k <= upper in
+    if List.for_all causal replay_steps
+    then if IdSet.mem proc_id pre_s.enabled
+         then Some replay_steps
+         else let is_parent k s = k > lower && k < time - upper && s.run.op = Spawn && s.run.obj_ptr = Some proc_id in
+              match list_findi is_parent state with
+              | None -> None
+              | Some (parent_i, spawn) ->
+                  assert (parent_i > lower) ;
+                  begin match find_replay_trace ~lower:parent_i ~upper spawn.run.proc_id with
+                  | None -> None
+                  | Some spawn_steps -> Some (spawn_steps @ replay_steps)
+                  end
+    else None
+  in
+  match find_loc proc with
+  | None -> ()
+  | Some (i, _) ->
+    assert (List.length state = time) ;
+    let pre_s = List.nth state (time - i + 1) in
+    match find_replay_trace ~lower:0 ~upper:i proc.proc_id with
+    | None -> ()
+    | Some replay_steps ->
+        if match IdMap.find_opt j pre_s.backtrack with
+        | None -> true
+        | Some lst -> List.length lst > List.length replay_steps
+        then pre_s.backtrack <- IdMap.add j replay_steps pre_s.backtrack
+
+let map_diff_set map set =
+  IdMap.filter (fun key _ -> not (IdSet.mem key set)) map
+
+let rec explore func time state (explored, state_planned) current_schedule clock (last_read, last_write) =
+  let s = List.hd state in
+  assert (IdMap.is_empty s.backtrack) ;
+  let dones = ref IdSet.empty in
+  begin match state_planned with
+  | [] ->
+    if IdSet.cardinal s.enabled > 0 then begin
+      let p = IdSet.min_elt s.enabled in
+      let init_step = IdMap.find p s.procs in
+      s.backtrack <- IdMap.singleton p [init_step] ;
+    end
+  | { proc_id ; _ } :: _ ->
+      dones := explored ;
+      s.backtrack <- IdMap.singleton proc_id state_planned
+  end ;
+  let is_backtracking = ref false in
+  while IdMap.(cardinal (map_diff_set s.backtrack !dones)) > 0 do
+    let j, new_steps = IdMap.min_binding (map_diff_set s.backtrack !dones) in
+    let new_explored =
+      if !is_backtracking || state_planned <> [] then !dones else IdSet.empty in
```

Comment:
I think that `If not backtracking && state_planned is empty`, then `!dones` is an empty set anyway.